### PR TITLE
Clean up warnings for 2023 data pipeline

### PR DESCRIPTION
**Pull request by grgmiller**
### Purpose
This PR addresses several warnings that were being raised in the 2023 data pipeline:
- When flagging differences between the inputs and outputs of the EIA-923 allocation process, update threshold to only warn if differences are greater than 0.01% to avoid warnings for rounding errors.
- Add a multi-year backstop for sulfur content so that we have complete sulfur content data for JF fuels
- Drop bad generators from the EIA-923 pipeline that were introduced by https://github.com/catalyst-cooperative/pudl/issues/3987
- Fix issue with missing fuel data for a plant that had recently retired, so wasn't appearing in the primary fuel table even though it was still reporting CEMS data
- Flag and drop CEMS units that report steam-only output and are not mapped to an EIA ID. These are likely for district steam systems and not power generation.
- Add the units flagged as "Manual CAMD Excluded" in the PSDC to the list of non-grid-connected units to remove.

I pulled in the manual mappings from the other PR that were not related to steam-only units without a clear mapping. 

Advances CAR-4691

### Testing
Running pipeline for 2023
Will run for other years as well

### Usage Example/Visuals
Example of warning about CEMS units missing EIA generator mappings (these are actually handled by our manual fuel category reference table since they do not report to EIA):
image

Example of steam units being removed:
image

### Review estimate
20 minutes

### Future work
We are still missing some records in the plant attribute table:
image

There are still some mis-allocations happening. Something that probably needs to be investigated in PUDL sometime in the future:
image

### Checklist
- [ ] Update the documentation to reflect changes made in this PR
- [x] Format all updated python files using `black`
- [x] Clear outputs from all notebooks modified
- [x] Add docstrings and type hints to any new functions created

### CODE REVIEW

**Comment by rouille** [comment]
The pipeline finished successfully. That sais in the monthly an annual power sector data, there are entries with missing `fuel category`. For these rows all columns are 0, e.g, see annual MISO power sector data below:
[MISO.csv](https://github.com/user-attachments/files/18050885/MISO.csv)

**Comment by grgmiller** [comment]
> The pipeline finished successfully. That sais in the monthly an annual power sector data, there are entries with missing fuel category. For these rows all columns are 0, e.g, see annual MISO power sector data below:
[MISO.csv](https://github.com/user-attachments/files/18050885/MISO.csv)

I'm not seeing any missing fuel categories in my files that I got from running my pipeline yesterday. I think I must have fixed this already with one of the last commits maybe?

File: src/oge/data_cleaning.py [hunk 1]
```diff
@@ -889,6 +1025,76 @@ def manually_remove_steam_units(df):
     return df
 
 
+def identify_and_remove_steam_only_units(cems, year):
```
Comment (rouille) [inline src/oge/data_cleaning.py:1028]: Missing docstring

File: src/oge/data_cleaning.py [hunk 2]
```diff
@@ -697,14 +807,16 @@ def remove_plants(
...
+        df = identify_and_remove_steam_only_units(df, year)
+
     if distribution_connected_plants:
         pass
 
     return df
 
 
-def remove_non_grid_connected_plants(df):
+def remove_non_grid_connected_plants(df, year):
```
Comment (rouille) [inline src/oge/data_cleaning.py:819]: Outdated docstring

File: src/oge/data_cleaning.py [hunk 3]
```diff
@@ -889,6 +1025,76 @@ def manually_remove_steam_units(df):
...
+            "These may need to be mapped to prevent double-counting of generation"
+        )
+        logger.warning(potential_missing_map.to_string())
+
+    # flag units that report zero generation (but non-zero fuel consumption or steam), but which have no mapping to an EIA generator
+    # we want to remove these so that we don't include these emissions in the plant
+    anomalous_units = annual_cems[
+        (annual_cems["generator_id"].isna())
+        & (annual_cems["gross_generation_mwh"] == 0)
+    ]
```
Comment (rouille) [inline src/oge/data_cleaning.py:1070]: I don't think this filter does what it is intended to do, i.e., select zero-generation and non-zero fuel consumption or steam

File: src/oge/data_cleaning.py [hunk 4]
```diff
@@ -1774,6 +1981,7 @@ def aggregate_subplant_data_to_fleet(
     combined_subplant_data: pd.DataFrame,
     plant_attributes_table: pd.DataFrame,
     primary_fuel_table: pd.DataFrame,
+    year: int,
 ) -> pd.DataFrame:
     """Group plant data by BA and fuel category (fleet).
```
Comment (rouille) [inline src/oge/data_cleaning.py:1987]: Missing `year` in docstring

File: src/oge/impute_hourly_profiles.py [hunk 5]
```diff
@@ -745,6 +747,7 @@ def identify_missing_profiles(
     residual_profiles: pd.DataFrame,
     plant_attributes: pd.DataFrame,
     primary_fuel_table: pd.DataFrame,
+    year,
 ) -> pd.DataFrame:
     """Identifies the fleet-months for which there is no profile data in residual_profiles
```
Comment (rouille) [inline src/oge/impute_hourly_profiles.py:752]: missing `year` in docstring

File: src/oge/impute_hourly_profiles.py [hunk 6]
```diff
@@ -1250,6 +1257,7 @@ def aggregate_eia_data_to_fleet(
     monthly_eia_data_to_shape: pd.DataFrame,
     plant_attributes: pd.DataFrame,
     primary_fuel_table: pd.DataFrame,
+    year: int,
 ) -> pd.DataFrame:
     """Aggregates monthly EIA-923 data to the fleet level prior to assigning an hourly
```
Comment (rouille) [inline src/oge/impute_hourly_profiles.py:1262]: Missing `year` in docstring

File: src/oge/validation.py [hunk 7]
```diff
@@ -437,22 +437,24 @@ def test_chp_allocation(df):
...
-def test_for_missing_energy_source_code(df):
-    """Checks that there are no missing energy source codes associated with non-zero fuel consumption."""
-    logger.info(
-        "Checking that there are no missing energy source codes associated with non-zero fuel consumption...  "
-    )
-    missing_esc_test = df[
-        (df["energy_source_code"].isna()) & (df["fuel_consumed_mmbtu"] > 0)
-    ]
+def test_for_missing_energy_source_code(df: pd.DataFrame, drop_missing: bool = False):
+    """Checks that there are no missing energy source codes"""
```
Comment (rouille) [inline src/oge/validation.py:441]: Incomplete docstring

File: src/oge/data_cleaning.py (hunk 2 above)
Comment (grgmiller) [reply src/oge/data_cleaning.py:819]: Updated

File: src/oge/data_cleaning.py (hunk 1 above)
Comment (grgmiller) [reply src/oge/data_cleaning.py:1028]: Added

File: src/oge/data_cleaning.py (hunk 3 above)
Comment (grgmiller) [reply src/oge/data_cleaning.py:1070]: Fixed, split this logic out into several functions

File: src/oge/data_cleaning.py (hunk 4 above)
Comment (grgmiller) [reply src/oge/data_cleaning.py:1987]: added

File: src/oge/impute_hourly_profiles.py (hunk 5 above)
Comment (grgmiller) [reply src/oge/impute_hourly_profiles.py:752]: added

File: src/oge/impute_hourly_profiles.py (hunk 6 above)
Comment (grgmiller) [reply src/oge/impute_hourly_profiles.py:1262]: added

File: src/oge/validation.py (hunk 7 above)
Comment (grgmiller) [reply src/oge/validation.py:441]: Updated

**Review by rouille — APPROVED** [review]
🚢

**Comment by rouille** [comment]
Pipeline stopped in Step 4 with commit https://github.com/singularity-energy/open-grid-emissions/pull/400/commits/48b4a0f5bf9678405383bc61e81a38c1c6675d64
```
2024-12-13 11:21:36 [INFO] oge.data_pipeline:209 4. Cleaning CEMS data
Traceback (most recent call last):
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/data_pipeline.py", line 704, in <module>
    main(sys.argv[1:])
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/data_pipeline.py", line 210, in main
    cems = data_cleaning.clean_cems(
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/data_cleaning.py", line 975, in clean_cems
    cems = remove_negative_cems_data(cems)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: remove_negative_cems_data() missing 1 required positional argument: 'year'
```

**Comment by rouille** [comment]
Another one:
```
Traceback (most recent call last):
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/data_pipeline.py", line 704, in <module>
    main(sys.argv[1:])
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/data_pipeline.py", line 210, in main
    cems = data_cleaning.clean_cems(
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/data_cleaning.py", line 981, in clean_cems
    cems = remove_plants(
           ^^^^^^^^^^^^^^
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/data_cleaning.py", line 811, in remove_plants
    df = remove_unmapped_fuel(df, year)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/data_cleaning.py", line 927, in remove_unmapped_fuel
    validation.limit_error_output_df(potential_missing_map.to_string())
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/validation.py", line 2292, in limit_error_output_df
    return pd.concat([df.head(100), df.tail(100)], axis=0)
                      ^^^^^^^
AttributeError: 'str' object has no attribute 'head'
```

**Comment by rouille** [comment]
Other one:
```
Traceback (most recent call last):
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/data_pipeline.py", line 704, in <module>
    main(sys.argv[1:])
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/data_pipeline.py", line 210, in main
    cems = data_cleaning.clean_cems(
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/data_cleaning.py", line 979, in clean_cems
    cems = remove_plants(
           ^^^^^^^^^^^^^^
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/data_cleaning.py", line 811, in remove_plants
    df = remove_unmapped_fuel(df, year)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/data_cleaning.py", line 948, in remove_unmapped_fuel
    validation.limit_error_output_df(fuel_only_unmapped.to_string(index=False))
  File "/Users/brdo/Singularity/open-grid-emissions/src/oge/validation.py", line 2292, in limit_error_output_df
    return pd.concat([df.head(100), df.tail(100)], axis=0)
                      ^^^^^^^
AttributeError: 'str' object has no attribute 'head'
```

**Pull request merged**